File: backend/src/main/kotlin/io/ticktag/persistence/ticket/dto/TicketFilter.kt
package io.ticktag.persistence.ticket.dto

import io.ticktag.persistence.project.entity.Project
import io.ticktag.persistence.ticket.entity.AssignedTicketUser
import io.ticktag.persistence.ticket.entity.Progress
import io.ticktag.persistence.ticket.entity.Ticket
import io.ticktag.persistence.ticket.entity.TicketTag
import io.ticktag.persistence.user.entity.User
import org.springframework.data.jpa.domain.Specification
import java.time.Instant
import java.util.*
import javax.persistence.criteria.*


data class TicketFilter(val project: UUID, val numbers: List<Int>?, val title: String?, val tags: List<String>?,
                        val users: List<String?>?, val progressOne: Float?, val progressTwo: Float?,
                        val progressGreater: Boolean?, val dueDateOne: Instant?, val dueDateTwo: Instant?,
                        val dueDateGreater: Boolean?, val storyPointsOne: Int?, val storyPointsTwo: Int?,
                        val storyPointsGreater: Boolean?, val open: Boolean?, val parent: Int?) : Specification<Ticket> {


    val predicates = emptyList<Predicate>().toMutableList()

    override fun toPredicate(root: Root<Ticket>, query: CriteriaQuery<*>, cb: CriteriaBuilder): Predicate? {

        predicates.add(cb.equal(root.get<Project>("project").get<UUID>("id"), project))

        if (numbers != null) {
            predicates.add(cb.isTrue(root.get<Int>("number").`in`(numbers)))
        }
        if (title != null) {
            predicates.add(cb.like(cb.lower(root.get<String>("title")), title.toLowerCase().split(' ').joinToString("%", "%", "%")))
        }
        if (tags != null) {
            val join = root.join<Ticket, TicketTag>("tags")
            val ticketTagPath = root.get<TicketTag>("tags")
            query.multiselect(ticketTagPath)
            query.groupBy(root.get<UUID>("id"), root.get<Progress>("progress").get<Float>("totalProgress"))
            query.having(cb.greaterThanOrEqualTo(cb.count(join.get<TicketTag>("normalizedName")), tags.size.toLong()))
            val ttags = join.get<TicketTag>("normalizedName")
            predicates.add(cb.isTrue(ttags.`in`(tags)))
        }
        if (users != null) {
            val join = root.join<Ticket, AssignedTicketUser>("assignedTicketUsers", JoinType.LEFT)
            val userPath = root.get<User>("assignedTicketUsers")
            query.multiselect(userPath)
            query.groupBy(root.get<UUID>("id"), join.get<Ticket>("ticket").get<UUID>("id"), root.get<Progress>("progress").get<Float>("totalProgress"))
            val tusers = join.get<User>("user").get<String>("username")
            if(!users.contains("none")){
                predicates.add(cb.isTrue(tusers.`in`(users)))
            }
            else {
                predicates.add(cb.isNull(join.get<User>("user")))
            }
        }
        if (progressOne != null) {
            if (progressTwo != null) {
                predicates.add(cb.between(root.get<Progress>("progress").get<Float>("totalProgress"), progressOne, progressTwo))
            } else if (progressGreater != null) {
                if (progressGreater == true) {
                    predicates.add(cb.greaterThan(root.get<Progress>("progress").get<Float>("totalProgress"), progressOne))
                } else {
                    predicates.add(cb.lessThan(root.get<Progress>("progress").get<Float>("totalProgress"), progressOne))
                }
            } else {
                predicates.add(cb.equal(root.get<Progress>("progress").get<List<String>>("totalProgress"), progressOne))
            }
        }
        if (dueDateOne != null) {
            if (dueDateTwo != null) {
                predicates.add(cb.between(root.get<Instant>("dueDate"), dueDateOne, dueDateTwo))
            } else if (dueDateGreater != null) {
                if (dueDateGreater == true) {
                    predicates.add(cb.greaterThan(root.get<Instant>("dueDate"), dueDateOne))
                } else if (dueDateGreater == false) {
                    predicates.add(cb.lessThan(root.get<Instant>("dueDate"), dueDateOne))
                }
            } else {
                predicates.add(cb.equal(root.get<Instant>("dueDate"), dueDateOne))
            }
        }
        if (storyPointsOne != null) {
            if (storyPointsTwo != null) {
                predicates.add(cb.between(root.get<Int>("storyPoints"), storyPointsOne, storyPointsTwo))
            } else if (storyPointsGreater != null) {
                if (storyPointsGreater == true) {
                    predicates.add(cb.greaterThan(root.get<Int>("storyPoints"), storyPointsOne))
                } else {
                    predicates.add(cb.lessThan(root.get<Int>("storyPoints"), storyPointsOne))
                }
            } else {
                predicates.add(cb.equal(root.get<Int>("storyPoints"), storyPointsOne))
            }
        }
        if (open != null) {
            predicates.add(cb.equal(root.get<Boolean>("open"), open))
        }
        if (parent != null) {
            val isChildOf: Predicate
            if (parent < 0) {
                isChildOf = cb.isNull(root.get<Ticket>("parentTicket"))
            } else {
                isChildOf = cb.equal(root.get<Ticket>("parentTicket").get<Int>("number"), parent)
            }
            predicates.add(isChildOf)
        }
        return if (predicates.size <= 0) {
            null
        } else {
            cb.and(*predicates.toTypedArray())
        }

    }

}

File: frontend/src/app/ui/ticket-overview/ticket-filter/ticket-filter.ts
import * as moment from 'moment';

export class TicketFilter {
    readonly title: string = '';
    readonly ticketNumbers: number[];
    readonly tags: string[];
    readonly users: string[];
    readonly progressOne: number;
    readonly progressTwo: number;
    readonly progressGreater: boolean;
    readonly dueDateOne: number;
    readonly dueDateTwo: number;
    readonly dueDateGreater: boolean;
    readonly storyPointsOne: number;
    readonly storyPointsTwo: number;
    readonly storyPointsGreater: boolean;
    readonly open: boolean;
    readonly parentNumber: number;


    constructor(title: string, ticketNumbers: number[], tags: string[], users: string[],
        progressOne: number, progressTwo: number, progressGreater: boolean,
        dueDateOne: number, dueDateTwo: number, dueDateGreater: boolean, storyPointsOne: number,
        storyPointsTwo: number, storyPointsGreater: boolean, open: boolean, parentNumber: number) {
        this.title = title === '' ? undefined : title;
        this.ticketNumbers = ticketNumbers ? ticketNumbers.filter(e => isFinite(e)) : undefined;
        this.tags = (tags ? tags : []).length > 0 ? tags : undefined;
        this.users = (users ? users : []).length > 0 ? users : undefined;
        this.progressOne = progressOne !== progressOne ? undefined : progressOne;
        this.progressTwo = progressTwo !== progressTwo ? undefined : progressTwo;
        this.progressGreater = progressGreater;
        this.dueDateOne = dueDateOne !== dueDateOne ? undefined : dueDateOne;
        this.dueDateTwo = dueDateTwo !== dueDateTwo ? undefined : dueDateTwo;
        this.dueDateGreater = dueDateGreater;
        this.storyPointsOne = storyPointsOne;
        this.storyPointsTwo = storyPointsTwo;
        this.storyPointsGreater = storyPointsGreater;
        this.open = open;
        this.parentNumber = parentNumber;
    }

    toTicketFilterString(): string {
        let list: string[] = [];
        if (this.title) { list.push(this.title); }
        if (this.ticketNumbers) { list.push('!#:' + this.ticketNumbers); }
        if (this.tags) { list.push('!tag:' + this.tags); }
        if (this.users) { list.push('!user:' + this.users); }
        if (this.progressOne && this.progressTwo) {
            list.push('!progress:' + this.progressOne + '-' + this.progressTwo);
        } else if (this.progressOne && this.progressGreater !== undefined) {
            list.push('!progress:' + (this.progressGreater ? '>' : '<') + this.progressOne);
        } else if (this.progressOne && this.progressGreater === undefined) {
            list.push('!progress:' + this.progressOne);
        }
        if (this.dueDateOne && this.dueDateTwo) {
            list.push(
                '!dueDate:' + moment(this.dueDateOne, 'x').format('YYYY-MM-DD') + '-' + moment(this.dueDateTwo, 'x').format('YYYY-MM-DD')
                );
        } else if (this.dueDateOne && this.dueDateGreater !== undefined) {
            list.push('!dueDate:' + (this.dueDateGreater ? '>' : '<') + moment(this.dueDateOne, 'x').format('YYYY-MM-DD'));
        } else if (this.dueDateOne && this.dueDateGreater === undefined) {
            list.push('!dueDate:' + moment(this.dueDateOne, 'x').format('YYYY-MM-DD'));
        }
        if (this.storyPointsOne && this.storyPointsTwo) {
            list.push('!sp:' + this.storyPointsOne + '-' + this.storyPointsTwo);
        } else if (this.storyPointsOne && this.storyPointsGreater !== undefined) {
            list.push('!sp:' + (this.storyPointsGreater ? '>' : '<') + this.storyPointsOne);
        } else if (this.storyPointsOne && this.storyPointsGreater === undefined) {
            list.push('!sp:' + this.storyPointsOne);
        }
        if (this.open !== undefined) { list.push('!open:' + this.open); }
        if (this.parentNumber !== undefined) { list.push('!parent:' + ((this.parentNumber < 0) ? 'none' : this.parentNumber)); }
        let joinedList = list.join(' ');
        if (joinedList.length > 0 && joinedList.charAt(0) === ' ') {
            joinedList = joinedList.substring(1);
        }
        return joinedList;
    }
    toTicketFilterURLString(): string {
        let list: string[] = [];
        if (this.title) { list.push('title=' + encodeURIComponent(this.title)); }
        if (this.ticketNumbers) { list.push('ticketNumber=' + this.ticketNumbers); }
        if (this.tags) { list.push('tag=' + this.tags.map(e => encodeURIComponent(e))); }
        if (this.users) { list.push('user=' + this.users.map(e => encodeURIComponent(e))); }
        if (this.progressOne && this.progressTwo) {
            list.push('progressOne=' + this.progressOne);
            list.push('progressTwo=' + this.progressTwo);
        } else if (this.progressOne && this.progressGreater !== undefined) {
            list.push('progressOne=' + this.progressOne);
            list.push('progressGreater=' + this.progressGreater);
        } else if (this.progressOne && this.progressGreater === undefined) {
            list.push('progressOne=' + this.progressOne);
        }
        if (this.dueDateOne && this.dueDateTwo) {
            list.push('dueDateOne=' + moment(this.dueDateOne, 'x').valueOf());
            list.push('dueDateTwo=' + moment(this.dueDateTwo, 'x').valueOf());
        } else if (this.dueDateOne && this.dueDateGreater !== undefined) {
            list.push('dueDateOne=' + moment(this.dueDateOne, 'x').valueOf());
            list.push('dueDateGreater=' + this.dueDateGreater);
        } else if (this.dueDateOne && this.dueDateGreater === undefined) {
            list.push('dueDateOne=' + moment(this.dueDateOne, 'x').valueOf());
        }
        if (this.storyPointsOne && this.storyPointsTwo) {
            list.push('spOne=' + this.storyPointsOne);
            list.push('spTwo=' + this.storyPointsTwo);
        } else if (this.storyPointsOne && this.storyPointsGreater !== undefined) {
            list.push('spOne=' + this.storyPointsOne);
            list.push('spGreater=' + this.storyPointsGreater);
        } else if (this.storyPointsOne && this.storyPointsGreater === undefined) {
            list.push('spOne=' + this.storyPointsOne);
        }

        if (this.open !== undefined) { list.push('open=' + this.open); }
        if (this.parentNumber !== undefined) { list.push('parent=' + this.parentNumber); }
        return list.join('&');
    }
}
